### ISSUE TITLE
[Bug]: "remove --unused" doesn't remove SDKs pulled in as deps

### ISSUE THREAD

**Issue by kparal** [flatpak/flatpak#5192]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.14.0

### What Linux distribution are you using?

Fedora Linux

### Linux distribution version

37

### What architecture are you using?

x86_64

### How to reproduce

1. Install `org.vim.Vim` from Flathub
2. See that `org.freedesktop.Sdk` is pulled in as a dependency
3. Remove `org.vim.Vim`, see that `org.freedesktop.Sdk`  is not removed
4. Run `flatpak remove --unused`, see that `org.freedesktop.Sdk` is not offered to be removed

### Expected Behavior

`flatpak remove --unused` should remove even SDKs, when they were pulled in as dependencies, not just Platforms.

### Actual Behavior

It seems that `flatpak remove --unused` ignores installed SDKs, even if they were pulled in just as a dependency.

### Additional Information

Here's a full action log. Vim pulls in an SDK:

```
$ flatpak install flathub org.vim.Vim
Looking for matches…
Required runtime for org.vim.Vim/x86_64/stable (runtime/org.freedesktop.Sdk/x86_64/22.08) found in remote flathub
Do you want to install it? [Y/n]: 

org.vim.Vim permissions:
    ipc      network     x11     file access [1]     dbus access [2]

    [1] /tmp, /var/tmp, host
    [2] org.freedesktop.Flatpak


        ID                                   Branch          Op          Remote           Download
 1. [✓] org.freedesktop.Sdk.Locale           22.08           i           flathub           18,0 kB / 338,7 MB
 2. [✓] org.freedesktop.Sdk                  22.08           i           flathub          186,2 MB / 512,0 MB
 3. [✓] org.vim.Vim                          stable          i           flathub           20,2 MB / 14,2 MB

Installation complete.
```

Removing Vim doesn't remove the SDK:
```
$ flatpak remove org.vim.Vim

        ID               Branch      Op
 1. [-] org.vim.Vim      stable      r

Uninstall complete.
```

Nor does the `--unused` switch:
```
$ flatpak remove --unused

Nothing unused to uninstall
```

**Comment by mwleeds** [comment]
This happens when you have some app installed that uses the corresponding Platform runtime, in this case `org.freedesktop.Platform/x86_64/22.08`, or you have installed `org.freedesktop.Platform/x86_64/22.08` directly (and therefore it's pinned). The logic is here:
https://github.com/flatpak/flatpak/blob/7a99c82900ea0e529335efae9e5210fe3366de46/common/flatpak-dir.c#L16740-L16747

So while this behavior is surprising when you don't know about that logic, it is the expected behavior. I think it would make sense to at least add some debug output for when `flatpak -v` is used; I'm not sure we want to add a message to explain what's going on to the user without `-v`.

**Comment by mwleeds** [comment]
#5237 should make things like this easier to debug in the future (when `flatpak -vv` is used) but I don't think we need to leave this issue open until that PR is merged.

**Comment by kparal** [comment]
Perhaps a silly question, but why isn't the SDK removed when it's not pinned? When I install an SDK manually, it gets pinned, which means it doesn't get auto-removed. That's correct, because I use that SDK for development. For example:

```
$ flatpak remove --unused

These runtimes in installation 'system' are pinned and won't be removed; see flatpak-pin(1):
  runtime/org.gnome.Sdk/x86_64/43
Nothing unused to uninstall
```

But when an SDK is pulled as a dependency for Vim, it's not pinned. Why isn't it safe to remove such an SDK (after Vim is gone)? Isn't that code snippet you shared simply outdated, i.e. written before platform/sdk pinning was a thing?

**Comment by rusty-snake** [comment]
Because vim (and other IDEs)¹ use the Sdk as their runtime too:

https://github.com/flathub/org.vim.Vim/blob/25c8dc28037b71c86b3ff0e59ac09da1f9b8c8f8/org.vim.Vim.json#L3-L5

and leaving vim without runtime isn't safe.

¹ Texteditor? IDE? Vim?

**Comment by kparal** [comment]
> Because vim (and other IDEs)¹ use the Sdk as their runtime too:

Sorry, one of is confused, I think 🙂 In the original description I showed that Vim pulled the SDK during install and then it wasn't removed when Vim was removed. So there was clearly no other app using that SDK and Vim wasn't using it either (because it was removed).

**Comment by mwleeds** [comment]
> @mwleeds Perhaps a silly question, but why isn't the SDK removed when it's not pinned? When I install an SDK manually, it gets pinned, which means it doesn't get auto-removed. That's correct, because I use that SDK for development. For example:
> 
> ```
> $ flatpak remove --unused
> 
> These runtimes in installation 'system' are pinned and won't be removed; see flatpak-pin(1):
>   runtime/org.gnome.Sdk/x86_64/43
> Nothing unused to uninstall
> ```
> 
> But when an SDK is pulled as a dependency for Vim, it's not pinned. Why isn't it safe to remove such an SDK (after Vim is gone)? Isn't that code snippet you shared simply outdated, i.e. written before platform/sdk pinning was a thing?

You're right, that logic is from before pinning was possible. And I see what you're saying, if you use that sdk for development you probably should pin it manually.

But if we start considering such sdk runtimes unused, isn't there a risk of causing disruption to users who might currently depend on using an sdk that's not pinned? Perhaps a rare scenario but it is the type of breaking change that may cause frustration.

**Comment by kparal** [comment]
> But if we start considering such sdk runtimes unused, isn't there a risk of causing disruption to users who might currently depend on using an sdk that's not pinned?

Pinning is available for some time already, right? And people are likely to use recent SDKs for development, rather than very old ones, I assume. But yes, it might happen. It's trivial to fix it though (just install it again - the error messages e.g. from flatpak-builder are very clear about what happened and what to do), so I don't think it's a problem, honestly. The developers can surely cope with this better than regular users who need to free disk space and don't know how 🙂
